library/std/src/sys/sync/condvar/queue.rs

```diff
@@ -0,0 +1,380 @@
+//! A generic `Condvar` implementation based on thread parking and a lockless
+//! queue of threads.
+//!
+//! Not all platforms provide an efficient `Condvar` implementation: the UNIX
+//! `pthread_condvar_t` needs memory allocation, while SGX doesn't have
+//! synchronization primitives at all. Therefore, we implement our own.
+//!
+//! To do so, we keep a list of the [`Thread`]s waiting on the `Condvar` and
+//! wake them up as needed. Access to the list is controlled by an atomic
+//! counter. To notify a waiter, the counter is incremented. If the counter
+//! was previously zero, the notifying thread has control over the list and
+//! will wake up threads until the number of threads it has woken up equals
+//! the counter value. Therefore, other threads do not need to wait for control
+//! over the list because the controlling thread will take over their notification.
+//!
+//! This counter is embedded into the lower bits of a pointer to the list head.
+//! As that limits the number of in-flight notifications, the counter increments
+//! are saturated to a maximum value ([`ALL`]) that causes all threads to be woken
+//! up, leading to a spurious wakeup for some threads. The API of `Condvar` permits
+//! this however. Timeouts employ the same method to make sure that the current
+//! thread handle is removed from the list.
+//!
+//! The list itself has the same structure as the one used by the queue-based
+//! `RwLock` implementation, see its documentation for more information. This
+//! enables the lockless enqueuing of threads and results in `Condvar` being
+//! only a pointer in size.
+//!
+//! This implementation is loosely based upon the lockless `Condvar` in
+//! [`usync`](https://github.com/kprotty/usync/blob/8937bb77963f6bf9068e56ad46133e933eb79974/src/condvar.rs).
```

Comment:
This worries me, since that repository is only MIT licensed -- and std needs to also be Apache licensed.

Reply:
This implementation shares no code with `usync` and contains significant alterations to the algorithm, so I don't think that this is a problem. Maybe "inspired by" is a better way to put it.

CC @kprotty what do you think?